In [ ]:
import pandas as pd
df = pd.read_csv('/content/Bengaluru_House_Data.csv')

In [ ]:
df.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [ ]:
df['society'].isnull().sum()

np.int64(5502)

In [ ]:
df.shape

(13320, 9)

In [ ]:
df = df.drop('society', axis=1)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   area_type     13320 non-null  object 
 1   availability  13320 non-null  object 
 2   location      13319 non-null  object 
 3   size          13304 non-null  object 
 4   total_sqft    13320 non-null  object 
 5   bath          13247 non-null  float64
 6   balcony       12711 non-null  float64
 7   price         13320 non-null  float64
dtypes: float64(3), object(5)
memory usage: 832.6+ KB


In [ ]:
df.isnull().sum()

area_type         0
availability      0
location          1
size             16
total_sqft        0
bath             73
balcony         609
price             0
dtype: int64

In [ ]:
df.dropna(subset=['location', 'size'], inplace=True)

In [ ]:
df.isnull().sum()

area_type         0
availability      0
location          0
size              0
total_sqft        0
bath             57
balcony         593
price             0
dtype: int64

In [ ]:
df['size'].unique()

array(['2 BHK', '4 Bedroom', '3 BHK', '4 BHK', '6 Bedroom', '3 Bedroom',
       '1 BHK', '1 RK', '1 Bedroom', '8 Bedroom', '2 Bedroom',
       '7 Bedroom', '5 BHK', '7 BHK', '6 BHK', '5 Bedroom', '11 BHK',
       '9 BHK', '9 Bedroom', '27 BHK', '10 Bedroom', '11 Bedroom',
       '10 BHK', '19 BHK', '16 BHK', '43 Bedroom', '14 BHK', '8 BHK',
       '12 Bedroom', '13 BHK', '18 Bedroom'], dtype=object)

In [ ]:
num_1rk = df[df['size'] == '1 RK'].shape[0]
print(f"Number of records for '1 RK': {num_1rk}")

Number of records for '1 RK': 13


# Task
Clean the dataframe by dropping the 'society' column, removing rows where 'location' and 'size' are null, creating 'BHK' and 'bedroom' columns based on the 'size' column, filling missing values in the new columns with 0, and removing rows where 'size' is '1 RK'.

## Create 'bhk' and 'bedroom' columns

### Subtask:
Extract the numerical part from the 'size' column and use it to populate the new 'BHK' and 'bedroom' columns based on whether the size string contains 'BHK' or 'Bedroom'.


In [ ]:
df['bhk'] = df['size'].apply(lambda x: int(x.split(' ')[0]) if 'BHK' in x else pd.NA)
df['bedroom'] = df['size'].apply(lambda x: int(x.split(' ')[0]) if 'Bedroom' in x else pd.NA)

## Handle '1 rk' values

### Subtask:
Remove rows where the 'size' column is '1 RK'.


**Reasoning**:
Filter the DataFrame to remove rows where the 'size' is '1 RK'.



In [ ]:
df = df[df['size'] != '1 RK']

## Fill missing values

### Subtask:
Fill any missing values in the newly created 'BHK' and 'bedroom' columns with 0.


**Reasoning**:
Fill the missing values in the 'bhk' and 'bedroom' columns with 0.



In [ ]:
df['bhk'] = df['bhk'].fillna(0)
df['bedroom'] = df['bedroom'].fillna(0)

/tmp/ipython-input-329339214.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['bhk'] = df['bhk'].fillna(0)
/tmp/ipython-input-329339214.py:2: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['bedroom'] = df['bedroom'].fillna(0)


In [ ]:
df.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

  total_sqft  bath  balcony   price  bhk  bedroom  
0       1056   2.0      1.0   39.07    2        0  
1       2600   5.0      3.0  120.00    0        4  
2       1440   2.0      3.0   62.00    3        0  
3       1521   3.0      1.0   95.00    3        0  
4       1200   2.0      1.0   51.00    2        0

In [ ]:
df['availability'].unique()

array(['19-Dec', 'Ready To Move', '18-May', '18-Feb', '20-Dec', '17-Oct',
       '21-Dec', '19-Sep', '20-Sep', '18-Mar', '20-Feb', '18-Apr',
       '20-Aug', '18-Oct', '19-Mar', '17-Sep', '18-Dec', '17-Aug',
       '19-Apr', '18-Jun', '22-Dec', '22-Jan', '18-Aug', '19-Jan',
       '17-Jul', '18-Jul', '21-Jun', '20-May', '19-Aug', '18-Sep',
       '17-May', '17-Jun', '21-May', '18-Jan', '20-Mar', '17-Dec',
       '16-Mar', '19-Jun', '22-Jun', '19-Jul', '21-Feb', '19-May',
       '17-Nov', '20-Oct', '20-Jun', '18-Nov', '19-Feb', '21-Oct',
       '21-Jan', '17-Mar', '17-Apr', '22-May', '19-Oct', '21-Jul',
       '21-Mar', '16-Dec', '22-Mar', '20-Jan', '21-Sep', '21-Aug',
       '14-Nov', '19-Nov', '15-Nov', '16-Jul', '15-Jun', '17-Feb',
       '20-Nov', '20-Jul', '16-Sep', '15-Oct', '21-Nov', '15-Dec',
       '16-Oct', '22-Nov', '15-Aug', '17-Jan', '16-Nov', '20-Apr',
       '16-Jan', '14-Jul'], dtype=object)

In [ ]:
df['availability'] = df['availability'].apply(lambda x: 'Ready To Move' if x == 'Ready To Move' else 'takes time to be available')

In [ ]:
df.head()

area_type                availability                  location  \
0  Super built-up  Area  takes time to be available  Electronic City Phase II   
1            Plot  Area               Ready To Move          Chikka Tirupathi   
2        Built-up  Area               Ready To Move               Uttarahalli   
3  Super built-up  Area               Ready To Move        Lingadheeranahalli   
4  Super built-up  Area               Ready To Move                  Kothanur   

        size total_sqft  bath  balcony   price  bhk  bedroom  
0      2 BHK       1056   2.0      1.0   39.07    2        0  
1  4 Bedroom       2600   5.0      3.0  120.00    0        4  
2      3 BHK       1440   2.0      3.0   62.00    3        0  
3      3 BHK       1521   3.0      1.0   95.00    3        0  
4      2 BHK       1200   2.0      1.0   51.00    2        0

In [ ]:
df['area_type'].unique()

array(['Super built-up  Area', 'Plot  Area', 'Built-up  Area',
       'Carpet  Area'], dtype=object)

In [ ]:
df.sample(10)

area_type                availability  \
7054   Super built-up  Area               Ready To Move   
2999   Super built-up  Area               Ready To Move   
9726             Plot  Area  takes time to be available   
2612   Super built-up  Area  takes time to be available   
7020         Built-up  Area               Ready To Move   
10388  Super built-up  Area               Ready To Move   
7101   Super built-up  Area               Ready To Move   
9952             Plot  Area               Ready To Move   
7070   Super built-up  Area               Ready To Move   
11758  Super built-up  Area               Ready To Move   

                       location       size total_sqft  bath  balcony   price  \
7054             Sarjapur  Road      2 BHK       1229   2.0      2.0   70.00   
2999         7th Phase JP Nagar      3 BHK       1850   3.0      2.0  140.00   
9726             Kanakpura Road  4 Bedroom       3170   4.0      3.0  250.00   
2612   Electronic City Phase II      3 BHK       1395   2.0      1.0   62.78   
7020               Indira Nagar      3 BHK       1740   3.0      2.0  191.00   
10388                  Cox Town      2 BHK       1000   2.0      3.0   58.00   
7101                 Kudlu Gate      3 BHK       1656   3.0      2.0   80.00   
9952              Chikkabanavar  4 Bedroom       1200   4.0      2.0  130.00   
7070                    Kengeri      1 BHK        400   2.0      1.0   25.00   
11758      5th Block Hbr Layout      5 BHK       1200   5.0      NaN  205.00   

       bhk  bedroom  
7054     2        0  
2999     3        0  
9726     0        4  
2612     3        0  
7020     3        0  
10388    2        0  
7101     3        0  
9952     0        4  
7070     1        0  
11758    5        0

In [ ]:
df[df['area_type'] == 'Plot  Area']

area_type                availability               location  \
1      Plot  Area               Ready To Move       Chikka Tirupathi   
9      Plot  Area               Ready To Move           Gandhi Bazar   
11     Plot  Area               Ready To Move             Whitefield   
14     Plot  Area               Ready To Move               Sarjapur   
22     Plot  Area               Ready To Move            Thanisandra   
...           ...                         ...                    ...   
13291  Plot  Area  takes time to be available         Weavers Colony   
13300  Plot  Area               Ready To Move          Hosakerehalli   
13303  Plot  Area               Ready To Move         Vidyaranyapura   
13306  Plot  Area               Ready To Move  Rajarajeshwari Nagara   
13311  Plot  Area               Ready To Move       Ramamurthy Nagar   

            size total_sqft  bath  balcony  price  bhk  bedroom  
1      4 Bedroom       2600   5.0      3.0  120.0    0        4  
9      6 Bedroom       1020   6.0      NaN  370.0    0        6  
11     4 Bedroom       2785   5.0      3.0  295.0    0        4  
14     3 Bedroom       2250   3.0      2.0  148.0    0        3  
22     4 Bedroom       2800   5.0      2.0  380.0    0        4  
...          ...        ...   ...      ...    ...  ...      ...  
13291  1 Bedroom        812   1.0      0.0   26.0    0        1  
13300  5 Bedroom       1500   6.0      2.0  145.0    0        5  
13303  5 Bedroom        774   5.0      3.0   70.0    0        5  
13306  4 Bedroom       1200   5.0      NaN  325.0    0        4  
13311  7 Bedroom       1500   9.0      2.0  250.0    0        7  

[2009 rows x 10 columns]

In [ ]:
houses = df[(df['area_type'] == 'Plot  Area') & (df['bedroom'] != 0)]

In [ ]:
houses.head()

area_type   availability          location       size total_sqft  bath  \
1   Plot  Area  Ready To Move  Chikka Tirupathi  4 Bedroom       2600   5.0   
9   Plot  Area  Ready To Move      Gandhi Bazar  6 Bedroom       1020   6.0   
11  Plot  Area  Ready To Move        Whitefield  4 Bedroom       2785   5.0   
14  Plot  Area  Ready To Move          Sarjapur  3 Bedroom       2250   3.0   
22  Plot  Area  Ready To Move       Thanisandra  4 Bedroom       2800   5.0   

    balcony  price  bhk  bedroom  
1       3.0  120.0    0        4  
9       NaN  370.0    0        6  
11      3.0  295.0    0        4  
14      2.0  148.0    0        3  
22      2.0  380.0    0        4

In [ ]:
houses[houses['bhk'] != 0]

Empty DataFrame
Columns: [area_type, availability, location, size, total_sqft, bath, balcony, price, bhk, bedroom]
Index: []

In [ ]:
houses = houses.drop('bhk', axis=1)

In [ ]:
plot_area_zero_bedroom = df[(df['area_type'] == 'Plot  Area') & (df['bedroom'] == 0)]
if plot_area_zero_bedroom.empty:
    print("There are no records in 'df' where 'area_type' is 'Plot Area' and 'bedroom' is 0.")
else:
    print(f"There are {len(plot_area_zero_bedroom)} records in 'df' where 'area_type' is 'Plot Area' and 'bedroom' is 0.")
    display(plot_area_zero_bedroom.head())

There are 1 records in 'df' where 'area_type' is 'Plot Area' and 'bedroom' is 0.


area_type   availability  location   size total_sqft  bath  balcony  \
7947  Plot  Area  Ready To Move  JP Nagar  3 BHK      20000   3.0      NaN   

      price  bhk  bedroom  
7947  175.0    3        0

In [ ]:
record_to_modify = df[(df['area_type'] == 'Plot  Area') & (df['bedroom'] == 0)]

In [ ]:
# Get the index of the record to modify
index_to_modify = record_to_modify.index[0]

# Update the 'bhk' and 'bedroom' values in the original DataFrame 'df'
df.loc[index_to_modify, 'bhk'] = 3
df.loc[index_to_modify, 'bedroom'] = 3

In [ ]:
houses = df[df['area_type'] == 'Plot  Area'].copy()

In [ ]:
display(houses.head())

area_type   availability          location       size total_sqft  bath  \
1   Plot  Area  Ready To Move  Chikka Tirupathi  4 Bedroom       2600   5.0   
9   Plot  Area  Ready To Move      Gandhi Bazar  6 Bedroom       1020   6.0   
11  Plot  Area  Ready To Move        Whitefield  4 Bedroom       2785   5.0   
14  Plot  Area  Ready To Move          Sarjapur  3 Bedroom       2250   3.0   
22  Plot  Area  Ready To Move       Thanisandra  4 Bedroom       2800   5.0   

    balcony  price  bhk  bedroom  
1       3.0  120.0    0        4  
9       NaN  370.0    0        6  
11      3.0  295.0    0        4  
14      2.0  148.0    0        3  
22      2.0  380.0    0        4

In [ ]:
houses = houses.drop(['bhk', 'size'], axis=1)

In [ ]:
houses.isnull().sum()

area_type         0
availability      0
location          0
total_sqft        0
bath              0
balcony         172
price             0
bedroom           0
dtype: int64

In [ ]:
houses.shape

(2009, 8)

In [ ]:
built_up_area_df = df[df['area_type'] == 'Built-up  Area']
display(built_up_area_df.head())

area_type                availability       location   size  \
2   Built-up  Area               Ready To Move    Uttarahalli  3 BHK   
13  Built-up  Area               Ready To Move      Gottigere  2 BHK   
20  Built-up  Area               Ready To Move        Kengeri  1 BHK   
27  Built-up  Area  takes time to be available     Whitefield  3 BHK   
34  Built-up  Area               Ready To Move  Kasturi Nagar  3 BHK   

   total_sqft  bath  balcony  price  bhk  bedroom  
2        1440   2.0      3.0   62.0    3        0  
13       1100   2.0      2.0   40.0    2        0  
20        600   1.0      1.0   15.0    1        0  
27       1610   3.0      2.0   81.0    3        0  
34       1925   3.0      NaN  125.0    3        0

# Task
Check if there are any rows in the dataframe `df` where the `area_type` is 'Plot Area' that are not present in the dataframe `houses`.

## Filter df for plot area

### Subtask:
Create a temporary DataFrame containing only rows from `df` where 'area_type' is 'Plot Area'.


In [ ]:
df_plot_area = df[df['area_type'] == 'Plot  Area']
display(df_plot_area.head())

area_type   availability          location       size total_sqft  bath  \
1   Plot  Area  Ready To Move  Chikka Tirupathi  4 Bedroom       2600   5.0   
9   Plot  Area  Ready To Move      Gandhi Bazar  6 Bedroom       1020   6.0   
11  Plot  Area  Ready To Move        Whitefield  4 Bedroom       2785   5.0   
14  Plot  Area  Ready To Move          Sarjapur  3 Bedroom       2250   3.0   
22  Plot  Area  Ready To Move       Thanisandra  4 Bedroom       2800   5.0   

    balcony  price  bhk  bedroom  
1       3.0  120.0    0        4  
9       NaN  370.0    0        6  
11      3.0  295.0    0        4  
14      2.0  148.0    0        3  
22      2.0  380.0    0        4

**Reasoning**:
Compare the temporary DataFrame `df_plot_area` with the `houses` DataFrame to identify rows in `df_plot_area` that are not present in `houses`.



In [ ]:
missing_rows = df_plot_area[~df_plot_area.index.isin(houses.index)]
if missing_rows.empty:
    print("There are no rows in 'df' where 'area_type' is 'Plot Area' that are not present in 'houses'.")
else:
    print(f"There are {len(missing_rows)} rows in 'df' where 'area_type' is 'Plot Area' that are not present in 'houses'.")
    display(missing_rows)

There are no rows in 'df' where 'area_type' is 'Plot Area' that are not present in 'houses'.


In [ ]:
df = df[df['area_type'] != 'Plot  Area']

In [ ]:
apartments = df

In [ ]:
apartments.shape

(11281, 10)

In [ ]:
houses.head()

area_type   availability          location total_sqft  bath  balcony  \
1   Plot  Area  Ready To Move  Chikka Tirupathi       2600   5.0      3.0   
9   Plot  Area  Ready To Move      Gandhi Bazar       1020   6.0      NaN   
11  Plot  Area  Ready To Move        Whitefield       2785   5.0      3.0   
14  Plot  Area  Ready To Move          Sarjapur       2250   3.0      2.0   
22  Plot  Area  Ready To Move       Thanisandra       2800   5.0      2.0   

    price  bedroom  
1   120.0        4  
9   370.0        6  
11  295.0        4  
14  148.0        3  
22  380.0        4

In [ ]:
houses[houses['balcony']==0]

area_type                availability  \
43     Plot  Area               Ready To Move   
68     Plot  Area               Ready To Move   
129    Plot  Area               Ready To Move   
133    Plot  Area               Ready To Move   
159    Plot  Area               Ready To Move   
...           ...                         ...   
13190  Plot  Area  takes time to be available   
13198  Plot  Area               Ready To Move   
13215  Plot  Area               Ready To Move   
13249  Plot  Area               Ready To Move   
13291  Plot  Area  takes time to be available   

                                       location total_sqft  bath  balcony  \
43                                    Yelahanka        660   1.0      0.0   
68                          Devarachikkanahalli       1350   7.0      0.0   
129                          Vishwapriya Layout        950   7.0      0.0   
133    Ramaswamy Palya - Kammanahalli Main Road       1200   4.0      0.0   
159                          Mahalakshmi Layout       3750   4.0      0.0   
...                                         ...        ...   ...      ...   
13190                                 Yelahanka       1330   5.0      0.0   
13198                            Pattegarhpalya       1200   8.0      0.0   
13215                               Frazer Town        896   1.0      0.0   
13249                                  KR Puram       1200   2.0      0.0   
13291                            Weavers Colony        812   1.0      0.0   

       price  bedroom  
43      48.0        1  
68      85.0        8  
129    115.0        7  
133    210.0        4  
159    760.0        4  
...      ...      ...  
13190  210.0        5  
13198  140.0        8  
13215  100.0        1  
13249   75.0        2  
13291   26.0        1  

[371 rows x 8 columns]

In [ ]:
houses = houses.drop('area_type', axis=1)

In [ ]:
houses.head()

availability          location total_sqft  bath  balcony  price  bedroom
1   Ready To Move  Chikka Tirupathi       2600   5.0      3.0  120.0        4
9   Ready To Move      Gandhi Bazar       1020   6.0      NaN  370.0        6
11  Ready To Move        Whitefield       2785   5.0      3.0  295.0        4
14  Ready To Move          Sarjapur       2250   3.0      2.0  148.0        3
22  Ready To Move       Thanisandra       2800   5.0      2.0  380.0        4

In [ ]:
houses[houses['availability'] != 'Ready To Move']

availability            location total_sqft  bath  \
248    takes time to be available          Meenakunte       4050   3.0   
296    takes time to be available           TC Palaya       1475   3.0   
339    takes time to be available           T.C PALYA       1200   2.0   
349    takes time to be available  Suragajakkanahalli         11   3.0   
382    takes time to be available              Anekal       1200   1.0   
...                           ...                 ...        ...   ...   
13033  takes time to be available          Whitefield       1500   3.0   
13085  takes time to be available          Whitefield       1500   3.0   
13145  takes time to be available         Bommasandra       1200   3.0   
13190  takes time to be available           Yelahanka       1330   5.0   
13291  takes time to be available      Weavers Colony        812   1.0   

       balcony  price  bedroom  
248        2.0  280.0        3  
296        1.0   98.0        3  
339        0.0   79.0        2  
349        2.0   74.0        3  
382        1.0   36.0        2  
...        ...    ...      ...  
13033      1.0   61.9        3  
13085      1.0   71.0        3  
13145      2.0  110.0        3  
13190      0.0  210.0        5  
13291      0.0   26.0        1  

[219 rows x 7 columns]

In [ ]:
houses['balcony'].unique()

array([3., 0., 2., 1.])

In [ ]:
houses['balcony'] = houses['balcony'].fillna(0)

In [ ]:
houses.isnull().sum()

availability    0
location        0
total_sqft      0
bath            0
balcony         0
price           0
bedroom         0
dtype: int64

In [ ]:
houses.sample(10)

availability            location total_sqft  bath  balcony  price  \
10846  Ready To Move     Prashanth Nagar       1200   5.0      0.0  260.0   
8005   Ready To Move  Kumaraswami Layout        600   6.0      3.0  100.0   
4577   Ready To Move      Kanakpura Road       1900   3.0      1.0   99.0   
10320  Ready To Move       BTM 2nd Stage       2400   3.0      2.0  400.0   
470    Ready To Move      Vidyaranyapura       1440   2.0      0.0   95.0   
8660   Ready To Move           TC Palaya       1200   3.0      0.0   66.0   
6659   Ready To Move        Kundalahalli       2500   5.0      2.0  350.0   
5477   Ready To Move          Rayasandra        540   1.0      1.0   16.5   
12753  Ready To Move         OMBR Layout        600   3.0      0.0  140.0   
1939   Ready To Move           Kothannur        930   6.0      1.0  135.0   

       bedroom  price_per_bedroom  
10846        6          43.333333  
8005         6          16.666667  
4577         3          33.000000  
10320        4         100.000000  
470          2          47.500000  
8660         3          22.000000  
6659         4          87.500000  
5477         1          16.500000  
12753        5          28.000000  
1939         6          22.500000

In [ ]:
houses['price_per_bedroom'] = houses['price'] / houses['bedroom']
houses['total_sqft'] = pd.to_numeric(houses['total_sqft'], errors='coerce')
houses['price_per_sqft'] = houses['price'] / houses['total_sqft']

In [ ]:
houses.sample(10)

availability                  location  total_sqft  bath  \
12127  takes time to be available               Kaggalipura      2200.0   4.0   
10111               Ready To Move              Anjana Nagar      1035.0   5.0   
7459                Ready To Move          Ramamurthy Nagar      1350.0   5.0   
2270                Ready To Move                 TC Palaya      4800.0   3.0   
5641   takes time to be available  Electronic City Phase II      6000.0   4.0   
2950                Ready To Move     8th block Koramangala      2400.0   3.0   
2844                Ready To Move               Adityanagar      1260.0   1.0   
3982                Ready To Move               Anand nagar      1200.0   2.0   
9551                Ready To Move      2nd Block Hbr Layout       950.0   5.0   
3704                Ready To Move                Whitefield      3227.0   5.0   

       balcony  price  bedroom  price_per_bedroom  price_per_sqft  
12127      1.0  150.0        3          50.000000        0.068182  
10111      1.0  170.0        5          34.000000        0.164251  
7459       1.0  120.0        5          24.000000        0.088889  
2270       3.0  302.0        4          75.500000        0.062917  
5641       0.0  260.0        4          65.000000        0.043333  
2950       2.0  290.0        3          96.666667        0.120833  
2844       0.0   41.0        1          41.000000        0.032540  
3982       0.0   70.0        2          35.000000        0.058333  
9551       2.0   55.0        5          11.000000        0.057895  
3704       0.0  250.0        5          50.000000        0.077471

# Task
Encode the 'location' column in the `houses` dataframe by replacing the location values with the average price per location and bedroom.

## Calculate average price per location and bedroom

### Subtask:
Group the `houses` DataFrame by 'location' and 'bedroom', and calculate the mean price for each group.


**Reasoning**:
Group the houses DataFrame by 'location' and 'bedroom' and calculate the mean price for each group to prepare for encoding.



In [ ]:
location_bedroom_price = houses.groupby(['location', 'bedroom'])['price'].mean()
display(location_bedroom_price.head())

location              bedroom
 Devarachikkanahalli  2           83.0
                      3          160.0
 ittamadu             3          110.0
 south                3          480.0
1 Annasandrapalya     11         150.0
Name: price, dtype: float64

## Create a mapping dictionary

### Subtask:
Convert the grouped data (`location_bedroom_price`) into a dictionary that maps each location and bedroom combination to its average price.


**Reasoning**:
Convert the grouped data (`location_bedroom_price`) into a dictionary that maps each location and bedroom combination to its average price.



In [ ]:
location_bedroom_price_map = location_bedroom_price.to_dict()

## Apply the encoding

### Subtask:
Use the `location_bedroom_price_map` dictionary to create a new column named 'encoded_location_price' in the `houses` DataFrame, representing the encoded location price based on the location and bedroom combination.


**Reasoning**:
Apply the mapping dictionary to create the 'encoded_location_price' column in the houses DataFrame.



In [ ]:
houses['encoded_location_price'] = houses.apply(lambda row: location_bedroom_price_map.get((row['location'], row['bedroom']), 0), axis=1)
display(houses.head())

availability          location  total_sqft  bath  balcony  price  \
1   Ready To Move  Chikka Tirupathi      2600.0   5.0      3.0  120.0   
9   Ready To Move      Gandhi Bazar      1020.0   6.0      0.0  370.0   
11  Ready To Move        Whitefield      2785.0   5.0      3.0  295.0   
14  Ready To Move          Sarjapur      2250.0   3.0      2.0  148.0   
22  Ready To Move       Thanisandra      2800.0   5.0      2.0  380.0   

    bedroom  price_per_bedroom  price_per_sqft  encoded_location_price  
1         4          30.000000        0.046154              126.428571  
9         6          61.666667        0.362745              370.000000  
11        4          73.750000        0.105925              376.468085  
14        3          49.333333        0.065778              130.813000  
22        4          95.000000        0.135714              380.000000

## Handle locations/bedroom combinations not in the mapping

### Subtask:
Handle locations/bedroom combinations that were not present when calculating the averages (e.g., fill with a default value or the overall average price).


**Reasoning**:
Calculate the overall average price of the 'price' column in the houses DataFrame.



In [ ]:
overall_average_price = houses['price'].mean()
print(f"Overall average price: {overall_average_price}")

Overall average price: 209.36814833250372


**Reasoning**:
Identify rows where 'encoded_location_price' is 0 and replace the value with the overall average price.



In [ ]:
houses['encoded_location_price'] = houses['encoded_location_price'].replace(0, overall_average_price)

**Reasoning**:
Verify that there are no more rows with 'encoded_location_price' equal to 0.



In [ ]:
zeros_in_encoded_price = houses[houses['encoded_location_price'] == 0]
if zeros_in_encoded_price.empty:
    print("There are no rows with 'encoded_location_price' equal to 0.")
else:
    print(f"There are {len(zeros_in_encoded_price)} rows with 'encoded_location_price' equal to 0.")
    display(zeros_in_encoded_price.head())

There are no rows with 'encoded_location_price' equal to 0.


In [ ]:
houses = houses.drop('location', axis=1)

In [ ]:
houses.head()

availability  total_sqft  bath  balcony  price  bedroom  \
1   Ready To Move      2600.0   5.0      3.0  120.0        4   
9   Ready To Move      1020.0   6.0      0.0  370.0        6   
11  Ready To Move      2785.0   5.0      3.0  295.0        4   
14  Ready To Move      2250.0   3.0      2.0  148.0        3   
22  Ready To Move      2800.0   5.0      2.0  380.0        4   

    price_per_bedroom  price_per_sqft  encoded_location_price  
1           30.000000        0.046154              126.428571  
9           61.666667        0.362745              370.000000  
11          73.750000        0.105925              376.468085  
14          49.333333        0.065778              130.813000  
22          95.000000        0.135714              380.000000

In [ ]:
houses['availability'] = houses['availability'].map({'Ready To Move': 1, 'takes time to be available': 0})

In [ ]:
houses.head()

availability  total_sqft  bath  balcony  price  bedroom  \
1              1      2600.0   5.0      3.0  120.0        4   
9              1      1020.0   6.0      0.0  370.0        6   
11             1      2785.0   5.0      3.0  295.0        4   
14             1      2250.0   3.0      2.0  148.0        3   
22             1      2800.0   5.0      2.0  380.0        4   

    price_per_bedroom  price_per_sqft  encoded_location_price  
1           30.000000        0.046154              126.428571  
9           61.666667        0.362745              370.000000  
11          73.750000        0.105925              376.468085  
14          49.333333        0.065778              130.813000  
22          95.000000        0.135714              380.000000

In [ ]:
houses.to_csv('houses.csv', index=False)

In [ ]:
apartments.head()

area_type                availability                  location  \
0  Super built-up  Area  takes time to be available  Electronic City Phase II   
2        Built-up  Area               Ready To Move               Uttarahalli   
3  Super built-up  Area               Ready To Move        Lingadheeranahalli   
4  Super built-up  Area               Ready To Move                  Kothanur   
5  Super built-up  Area               Ready To Move                Whitefield   

    size total_sqft  bath  balcony  price  bhk  bedroom  
0  2 BHK       1056   2.0      1.0  39.07    2        0  
2  3 BHK       1440   2.0      3.0  62.00    3        0  
3  3 BHK       1521   3.0      1.0  95.00    3        0  
4  2 BHK       1200   2.0      1.0  51.00    2        0  
5  2 BHK       1170   2.0      1.0  38.00    2        0

In [ ]:
apartments[(apartments['bedroom'] != 0)]

area_type                availability               location  \
56     Built-up  Area  takes time to be available            Devanahalli   
64     Built-up  Area               Ready To Move           Bommanahalli   
79     Built-up  Area               Ready To Move            ISRO Layout   
81     Built-up  Area  takes time to be available            Hennur Road   
85     Built-up  Area               Ready To Move            Hegde Nagar   
...               ...                         ...                    ...   
13155  Built-up  Area               Ready To Move       Doddabommasandra   
13187  Built-up  Area               Ready To Move          Chikkabanavar   
13226  Built-up  Area               Ready To Move  Raja Rajeshwari Nagar   
13246  Built-up  Area               Ready To Move         Sarjapur  Road   
13315  Built-up  Area               Ready To Move             Whitefield   

            size   total_sqft  bath  balcony  price  bhk  bedroom  
56     4 Bedroom  3010 - 3410   NaN      NaN  192.0    0        4  
64     8 Bedroom         3000   8.0      0.0  140.0    0        8  
79     6 Bedroom         4400   6.0      3.0  250.0    0        6  
81     4 Bedroom  2957 - 3450   NaN      NaN  224.5    0        4  
85     6 Bedroom         3000   7.0      2.0  210.0    0        6  
...          ...          ...   ...      ...    ...  ...      ...  
13155  2 Bedroom         1600   2.0      1.0   90.0    0        2  
13187  8 Bedroom         4000   7.0      0.0  110.0    0        8  
13226  8 Bedroom         6000   8.0      3.0  215.0    0        8  
13246  4 Bedroom         2758   4.0      1.0  240.0    0        4  
13315  5 Bedroom         3453   4.0      0.0  231.0    0        5  

[517 rows x 10 columns]

In [ ]:
apartments[(apartments['bedroom'] != 0) & (apartments['area_type'] != 'Built-up  Area')]

area_type   availability              location       size  \
941    Super built-up  Area  Ready To Move            Whitefield  4 Bedroom   
1033           Carpet  Area  Ready To Move            Nagasandra  6 Bedroom   
3862   Super built-up  Area  Ready To Move                Jigani  4 Bedroom   
6862           Carpet  Area  Ready To Move      Doddabommasandra  8 Bedroom   
8413           Carpet  Area  Ready To Move             Kothannur  4 Bedroom   
10791          Carpet  Area  Ready To Move  2nd Stage Nagarbhavi  5 Bedroom   
11543          Carpet  Area  Ready To Move     Maruthi Sevanagar  4 Bedroom   

        total_sqft  bath  balcony  price  bhk  bedroom  
941    3606 - 5091   NaN      NaN  304.0    0        4  
1033          1500   5.0      1.0  130.0    0        6  
3862   3293 - 5314   NaN      NaN  226.0    0        4  
6862          3150   5.0      2.0  145.0    0        8  
8413          2000   4.0      2.0   95.0    0        4  
10791         4000   4.0      3.0  240.0    0        5  
11543         2100   3.0      0.0  210.0    0        4

In [ ]:
houses['total_sqft'].dtype

dtype('float64')

# Task
Build a decision tree regressor for the houses dataframe with 5 fold cross validation and and use hyper parameters as params = {
    'max_depth': [3, 5, 10, 15, 20],
    'min_samples_split': [2, 5, 10, 20],
    'min_samples_leaf': [1, 2, 4, 10],
    'max_leaf_nodes': [None, 10, 20, 50],
    'criterion': ['squared_error', 'friedman_mse']
}

then print mse, r2, mape, rmse, and all the metrics possible to evaluate the model

## Prepare data for modeling

### Subtask:
Define the features (X) and target variable (y) from the `houses` DataFrame.


**Reasoning**:
Define the features (X) and target variable (y) from the houses DataFrame.



In [ ]:
X = houses.drop(['price', 'price_per_sqft', 'price_per_bedroom'], axis=1)
y = houses['price']

## Split data into training and testing sets

### Subtask:
Split the data into training and testing sets to evaluate the model's performance on unseen data.


**Reasoning**:
Split the features and target variable into training and testing sets.



In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print(f"Shape of X_train: {X_train.shape}")
print(f"Shape of X_test: {X_test.shape}")
print(f"Shape of y_train: {y_train.shape}")
print(f"Shape of y_test: {y_test.shape}")

Shape of X_train: (1607, 6)
Shape of X_test: (402, 6)
Shape of y_train: (1607,)
Shape of y_test: (402,)


## Define the model and hyperparameter grid

### Subtask:
Initialize a Decision Tree Regressor model and define the hyperparameter grid for tuning.


**Reasoning**:
Initialize a Decision Tree Regressor model and define the hyperparameter grid for tuning.



In [ ]:
from sklearn.tree import DecisionTreeRegressor
model = DecisionTreeRegressor()
params = {
    'max_depth': [3, 5, 10, 15, 20],
    'min_samples_split': [2, 5, 10, 20],
    'min_samples_leaf': [1, 2, 4, 10],
    'max_leaf_nodes': [None, 10, 20, 50],
    'criterion': ['squared_error', 'friedman_mse']
}

## Perform gridsearchcv with cross-validation

### Subtask:
Use GridSearchCV to find the best hyperparameters using 5-fold cross-validation.


**Reasoning**:
Use GridSearchCV to find the best hyperparameters for the Decision Tree Regressor model using 5-fold cross-validation and negative mean squared error as the scoring metric.



In [ ]:
from sklearn.model_selection import GridSearchCV
grid_search = GridSearchCV(model, params, cv=5, scoring='neg_mean_squared_error')
grid_search.fit(X_train, y_train)
print("Best parameters found: ", grid_search.best_params_)
print("Best cross-validation score (negative MSE): ", grid_search.best_score_)

Best parameters found:  {'criterion': 'squared_error', 'max_depth': 10, 'max_leaf_nodes': 20, 'min_samples_leaf': 4, 'min_samples_split': 2}
Best cross-validation score (negative MSE):  -8177.485471047687


## Train the model with the best hyperparameters

### Subtask:
Train the Decision Tree Regressor model using the best hyperparameters found by GridSearchCV.


**Reasoning**:
Get the best estimator from the grid search results.



In [ ]:
best_model = grid_search.best_estimator_

**Reasoning**:
Verify the hyperparameters of the best model.



In [ ]:
print(best_model)

DecisionTreeRegressor(max_depth=10, max_leaf_nodes=20, min_samples_leaf=4)


## Evaluate the model

### Subtask:
Calculate and print various regression metrics (MSE, R2, RMSE, MAPE, etc.) to evaluate the performance of the trained model on the test set.


**Reasoning**:
Calculate and print various regression metrics (MSE, R2, RMSE, MAPE, etc.) to evaluate the performance of the trained model on the test set.



In [ ]:
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_percentage_error
import numpy as np
y_pred = best_model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
rmse = np.sqrt(mse)
mape = mean_absolute_percentage_error(y_test, y_pred)
print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared (R2) Score: {r2}")
print(f"Root Mean Squared Error (RMSE): {rmse}")
print(f"Mean Absolute Percentage Error (MAPE): {mape}")

Mean Squared Error (MSE): 7459.044047206627
R-squared (R2) Score: 0.8691212878088708
Root Mean Squared Error (RMSE): 86.36575737644304
Mean Absolute Percentage Error (MAPE): 0.2676103586358435


## Summary:

### Data Analysis Key Findings

*   The best hyperparameters found by GridSearchCV with 5-fold cross-validation were: `{'criterion': 'friedman_mse', 'max_depth': 15, 'max_leaf_nodes': None, 'min_samples_leaf': 1, 'min_samples_split': 5}`.
*   The best cross-validation score (negative MSE) achieved with these parameters was approximately -6161.79.
*   On the test set, the trained model with the best hyperparameters achieved a Mean Squared Error (MSE) of approximately 3149.27.
*   The R-squared (R2) score on the test set was approximately 0.945, indicating a strong fit of the model to the data.
*   The Root Mean Squared Error (RMSE) on the test set was approximately 56.12.
*   The Mean Absolute Percentage Error (MAPE) on the test set was approximately 0.076, suggesting an average prediction error of around 7.6%.

### Insights or Next Steps

*   The Decision Tree Regressor model with the identified hyperparameters demonstrates strong performance in predicting house prices based on the evaluation metrics.
*   Further analysis could involve exploring feature importance from the trained decision tree to understand which features contribute most to the price prediction.


In [ ]:
import pickle

# Specify the filename
filename = 'decision_tree_regressor_model.pkl'

# Save the model to a pickle file
with open(filename, 'wb') as f:
    pickle.dump(best_model, f)

print(f"Model successfully dumped to {filename}")

Model successfully dumped to decision_tree_regressor_model.pkl


## Summary:

### Data Analysis Key Findings

* The best hyperparameters found by GridSearchCV with 5-fold cross-validation were: `{'criterion': 'friedman_mse', 'max_depth': 15, 'max_leaf_nodes': None, 'min_samples_leaf': 1, 'min_samples_split': 5}`.
* The best cross-validation score (negative MSE) achieved with these parameters was approximately -6161.79.
* On the test set, the trained model with the best hyperparameters achieved a Mean Squared Error (MSE) of approximately 3149.27.
* The R-squared (R2) score on the test set was approximately 0.945, indicating a strong fit of the model to the data.
* The Root Mean Squared Error (RMSE) on the test set was approximately 56.12.
* The Mean Absolute Percentage Error (MAPE) on the test set was approximately 0.076, suggesting an average prediction error of around 7.6%.

### Insights or Next Steps

* The Decision Tree Regressor model with the identified hyperparameters demonstrates strong performance in predicting house prices based on the evaluation metrics.
* Further analysis could involve exploring feature importance from the trained decision tree to understand which features contribute most to the price prediction.

## Summary:

### Data Analysis Key Findings

* The best hyperparameters found by GridSearchCV with 5-fold cross-validation were: `{'criterion': 'squared_error', 'max_depth': 10, 'max_leaf_nodes': 20, 'min_samples_leaf': 4, 'min_samples_split': 2}`.
* The best cross-validation score (negative MSE) achieved with these parameters was approximately -8177.49.
* On the test set, the trained model with the best hyperparameters achieved a Mean Squared Error (MSE) of approximately 7459.04.
* The R-squared (R2) score on the test set was approximately 0.869, indicating a good fit of the model to the data.
* The Root Mean Squared Error (RMSE) on the test set was approximately 86.37.
* The Mean Absolute Percentage Error (MAPE) on the test set was approximately 0.268, suggesting an average prediction error of around 26.8%.

### Insights or Next Steps

* The Decision Tree Regressor model, trained without the 'price_per_sqft' and 'price_per_bedroom' features, still shows reasonable performance.
* You can now use this trained `best_model` for making predictions.
* Further analysis could involve comparing the performance of this model with the previous one (which included 'price_per_sqft' and 'price_per_bedroom') to understand the impact of those features.
* You could also explore other regression algorithms or try further feature engineering.

In [ ]:
import xgboost as xgb
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeRegressor

# Define the XGBoost Regressor model
model = xgb.XGBRegressor(random_state=42)

# Define the hyperparameter grid
param_grid = {
    'n_estimators': [100, 200, 300],             # fewer trees = faster training
    'learning_rate': [0.05, 0.1, 0.2],           # controls step size
    'max_depth': [3, 5, 7],                      # controls tree depth (avoid 10+)
    'min_child_weight': [1, 3],                  # regularization to prevent overfitting
    'subsample': [0.8, 1.0],                     # row sampling
    'colsample_bytree': [0.8, 1.0],              # feature sampling
    'gamma': [0, 0.1],                           # min loss reduction to split
    'reg_alpha': [0, 0.5],                       # L1 regularization
    'reg_lambda': [1, 1.5],                      # L2 regularization
    'objective': ['reg:squarederror'],           # regression objective
    'tree_method': ['hist'],                     # faster on CPU than 'auto'
    'eval_metric': ['mae']                       # use MAE instead of RMSE
}

In [ ]:
from sklearn.model_selection import GridSearchCV
grid_search = GridSearchCV(model, param_grid, cv=5, scoring='neg_mean_squared_error')
grid_search.fit(X_train, y_train)
print("Best parameters found: ", grid_search.best_params_)
print("Best cross-validation score (negative MSE): ", grid_search.best_score_)

Best parameters found:  {'colsample_bytree': 1.0, 'eval_metric': 'mae', 'gamma': 0, 'learning_rate': 0.05, 'max_depth': 3, 'min_child_weight': 3, 'n_estimators': 100, 'objective': 'reg:squarederror', 'reg_alpha': 0.5, 'reg_lambda': 1.5, 'subsample': 0.8, 'tree_method': 'hist'}
Best cross-validation score (negative MSE):  -8483.035249026552


In [ ]:
best_xgb_model = grid_search.best_estimator_

In [ ]:
print(best_xgb_model)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=1.0, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric='mae', feature_types=None,
             feature_weights=None, gamma=0, grow_policy=None,
             importance_type=None, interaction_constraints=None,
             learning_rate=0.05, max_bin=None, max_cat_threshold=None,
             max_cat_to_onehot=None, max_delta_step=None, max_depth=3,
             max_leaves=None, min_child_weight=3, missing=nan,
             monotone_constraints=None, multi_strategy=None, n_estimators=100,
             n_jobs=None, num_parallel_tree=None, ...)


## Evaluate the model

### Subtask:
Calculate and print various regression metrics (MSE, R2, RMSE, MAPE, etc.) to evaluate the performance of the trained model on the test set.

**Reasoning**:
Calculate and print various regression metrics (MSE, R2, RMSE, MAPE, etc.) to evaluate the performance of the trained model on the test set.

In [ ]:
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_percentage_error, mean_absolute_error
import numpy as np
y_pred_xgb = best_xgb_model.predict(X_test)
mse_xgb = mean_squared_error(y_test, y_pred_xgb)
r2_xgb = r2_score(y_test, y_pred_xgb)
rmse_xgb = np.sqrt(mse_xgb)
mape_xgb = mean_absolute_percentage_error(y_test, y_pred_xgb)
mae_xgb = mean_absolute_error(y_test, y_pred_xgb)

print(f"XGBoost Model Evaluation:")
print(f"Mean Squared Error (MSE): {mse_xgb}")
print(f"R-squared (R2) Score: {r2_xgb}")
print(f"Root Mean Squared Error (RMSE): {rmse_xgb}")
print(f"Mean Absolute Percentage Error (MAPE): {mape_xgb}")
print(f"Mean Absolute Error (MAE): {mae_xgb}")

XGBoost Model Evaluation:
Mean Squared Error (MSE): 6848.037139174832
R-squared (R2) Score: 0.8798422054971142
Root Mean Squared Error (RMSE): 82.75286786072633
Mean Absolute Percentage Error (MAPE): 0.2230210678596327
Mean Absolute Error (MAE): 37.906833863519324


## Summary:

### Data Analysis Key Findings

* The XGBoost Regressor model was built and tuned using GridSearchCV with 5-fold cross-validation and the specified hyperparameter grid.
* The best hyperparameters found by GridSearchCV were: [Insert best parameters here after execution].
* The best cross-validation score (negative MSE) achieved was: [Insert best score here after execution].
* On the test set, the trained model with the best hyperparameters achieved the following metrics:
    * Mean Squared Error (MSE): [Insert MSE here after execution]
    * R-squared (R2) Score: [Insert R2 here after execution]
    * Root Mean Squared Error (RMSE): [Insert RMSE here after execution]
    * Mean Absolute Percentage Error (MAPE): [Insert MAPE here after execution]
    * Mean Absolute Error (MAE): [Insert MAE here after execution]

### Insights or Next Steps

* [Provide insights based on the evaluation metrics and compare with the Decision Tree model if desired.]
* You can now use this trained `best_xgb_model` for making predictions.
* Further analysis could involve exploring feature importance from the trained XGBoost model.
* Consider saving this model to a pickle file for future use.

## Summary:

### Data Analysis Key Findings

* The XGBoost Regressor model was built and tuned using GridSearchCV with 5-fold cross-validation and the specified hyperparameter grid.
* The best hyperparameters found by GridSearchCV were: `{'colsample_bytree': 1.0, 'eval_metric': 'mae', 'gamma': 0, 'learning_rate': 0.05, 'max_depth': 3, 'min_child_weight': 3, 'n_estimators': 100, 'objective': 'reg:squarederror', 'reg_alpha': 0.5, 'reg_lambda': 1.5, 'subsample': 0.8, 'tree_method': 'hist'}`.
* The best cross-validation score (negative MSE) achieved was: -8483.04.
* On the test set, the trained model with the best hyperparameters achieved the following metrics:
  * Mean Squared Error (MSE): 6848.04
  * R-squared (R2) Score: 0.880
  * Root Mean Squared Error (RMSE): 82.75
  * Mean Absolute Percentage Error (MAPE): 0.223
  * Mean Absolute Error (MAE): 37.91

### Insights or Next Steps

* The XGBoost Regressor model with the tuned hyperparameters shows good performance in predicting house prices, with an R-squared score of approximately 0.88 and a Mean Absolute Error of around 37.91.
* This model appears to perform slightly better than the Decision Tree Regressor model we previously built based on the R2 and MAE metrics.
* You can now use this trained `best_xgb_model` for making predictions.
* Further analysis could involve exploring feature importance from the trained XGBoost model to understand which features contribute most to the price prediction.
* Consider saving this model to a pickle file for future use.

In [ ]:
import pickle
filename = 'best_xgb_model.pkl'
with open(filename, 'wb') as f:
    pickle.dump(best_xgb_model, f)
print(f"XGBoost model successfully dumped to {filename}")

XGBoost model successfully dumped to best_xgb_model.pkl


**Reasoning**:
Now that the XGBoost model has been saved to a pickle file, I can load it.



**Reasoning**:
Load the trained XGBoost model from the pickle file.



In [ ]:
import pickle
filename = 'best_xgb_model.pkl'
with open(filename, 'rb') as f:
    loaded_model = pickle.load(f)

**Reasoning**:
Verify the type of the loaded model to ensure it is an XGBoost Regressor.



In [ ]:
apartments.to_csv('apartments.csv', index=False)
print("apartments.csv saved successfully.")

apartments.csv saved successfully.
